Dshop/apps/products_catalogue/tests/test_api_persistent_cart.py

```diff
@@ -0,0 +1,179 @@
+
+@pytest.mark.django_db
+def test_add_relogin_get(tv_product):
+    client = APIClient()
+    user = User.objects.create_user(username='testuser', password='testpassword')
+    client.force_authenticate(user)
+    product = tv_product
+
+    data = {
+        'items': [ {'product_pk': product.pk, 'quantity': 10} ] 
+    }
+    client.force_authenticate(user=None)
+    client.force_authenticate(user)
+    response = client.post(reverse("api_cart"), data, format='json')
+    data = response.data
+    first_item = data["items"][0]
+    assert response.status_code == status.HTTP_201_CREATED
+    assert Decimal(data["total"]) ==  Decimal(first_item['subtotal']) == Decimal(tv_product.price * 10)  
+    assert data["count"] == first_item["quantity"] == 10 
+    assert first_item["product_name"] == product.name
+
+
+@pytest.mark.django_db
+def test_add_ten_and_get(ten_tv_products):
+    client = APIClient()
+    tvs = ten_tv_products
+    user = User.objects.create_user(username='testuser', password='testpassword')
+    client.force_authenticate(user)
+    quantities = [1, 1, 6, 8, 3, 4, 2, 26, 1, 10]
+    total_quantity, total_price = counts_to_price_quantity(tvs, quantities)
+
+    data = {   
+    'items': [
+            {'product_pk': product.pk, 'quantity': quantity}
+            for product, quantity in zip(ten_tv_products, quantities)
+        ]
+    }
+    response = client.post(reverse("api_cart"), data, format="json")
+    assert response.status_code == status.HTTP_201_CREATED
+    assert Decimal(response.data['total']) == Decimal(total_price)
+    assert response.data['count'] == total_quantity
+
+
+@pytest.mark.django_db
+def test_delete_ten(ten_tv_products):
+    client = APIClient()
+    tvs = ten_tv_products
+    user = User.objects.create_user(username='testuser', password='testpassword')
+    client.force_authenticate(user)
+    quantities = [1, 1, 6, 8, 3, 4, 2, 26, 1, 10]
+    total_quantity, total_price = counts_to_price_quantity(tvs, quantities)
+
+    data = {   
+    'items': [
+            {'product_pk': product.pk, 'quantity': quantity}
+            for product, quantity in zip(ten_tv_products, quantities)
+        ]
+    }
+    response = client.post(reverse("api_cart"), data, format="json")
+    assert response.status_code == status.HTTP_201_CREATED
+    response = client.post(reverse("api_cart"), {}, format="json")
+    assert response.status_code == status.HTTP_201_CREATED
+    assert_data_empty(response.data)
+
+
+@pytest.mark.django_db
+def test_get_non_unique_pks(tv_product):
+    client = APIClient()
+    user = User.objects.create_user(username='testuser', password='testpassword')
+    client.force_authenticate(user)
+
+    data = {
+        'items':[
+            {'product_pk': tv_product.pk, 'quantity': 2},
+            {'product_pk': tv_product.pk, 'quantity': 3}
+        ]
+    }
+    response = client.post(reverse("api_cart"), data, format="json")
+    assert response.status_code == status.HTTP_400_BAD_REQUEST
```

Comment:
Please also add assert for error text. This is good, but it can fail on anything, not only on the error you expect it to.

Reply:
done